### PR TITLE
Feature/objects 1027

### PR DESCRIPTION
**Pull request by andreassiegel**
### What changes were proposed in this pull request?
- adds request and user details validation
- returns _400 Bad Request_ responses in case of invalid data
- _401 Unauthorized_ is only returned if the service user credentials (HTTP Basic) do not match the configuration

In addition, the code is cleaned up and refactored:
- removed unnecessary configuration classes
- removed unnecessarily complex model and conversion structures
- added interfaces for service classes

The newly added interfaces could also be used for [user-details-stormpath](https://github.com/SMARTRACTECHNOLOGY/smartcosmos-user-details-stormpath) and [user-details-service](https://github.com/SMARTRACTECHNOLOGY/smartcosmos-user-details-service).
### How is this patch documented?
- Javadoc
- Code
- README
### How was this patch tested?
- JUnit
- manually in Postman
#### Depends On

Nothing, but like for https://github.com/SMARTRACTECHNOLOGY/smartcosmos-user-details-stormpath/pull/17, we have to ensure the auth-server can handle the _400 Bad Request_ responses.

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -54,6 +54,17 @@ POST /authenticate
...
 ----
 
+----
+400 Bad Request
+----
+[source,json]
+----
+{
+    "code": 1,
+    "message": "Unknown user bob"
```
Comment (wfschneider-rfid) [inline README.adoc]: The preferred way of handling this situation is to return something like "your credentials do not match what we have". Being specific about not finding a user or the password does not match allows for fishing and is not good practice.

File: src/main/java/net/smartcosmos/cluster/userdetails/DevKitUserDetailsService.java [hunk 2]
```diff
@@ -32,4 +38,10 @@ PasswordEncoder passwordEncoder() {
 
         return new BCryptPasswordEncoder();
     }
+
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/DevKitUserDetailsService.java]: We have an evolving stylistic difference here.  I would prefer things like this be added in `@Configuration` classes unless there is a compelling reason to have them in the main class. We need to be consistent across applications about the location of things that are dynamic.  It's not likely but in the future we may need to make the encode configurable from the YML file, having it in a configuration class makes that easier.

File: src/main/java/net/smartcosmos/cluster/userdetails/domain/AuthenticateUserRequest.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package net.smartcosmos.cluster.userdetails.dto;
+package net.smartcosmos.cluster.userdetails.domain;
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/domain/AuthenticateUserRequest.java]: much nicer name

File: src/main/java/net/smartcosmos/cluster/userdetails/resource/AuthenticationResource.java [hunk 4]
```diff
@@ -9,32 +9,31 @@
...
 
-    private final AuthenticationService authenticationService;
+    private final AuthenticateUserService service;
 
     @Autowired
-    public AuthenticationResource(AuthenticationService authenticationService) {
+    public AuthenticationResource(AuthenticateUserService authenticateUserService) {
 
-        this.authenticationService = authenticationService;
+        this.service = authenticateUserService;
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/resource/AuthenticationResource.java]: technically if the variable names don't match then `this` is not required to clarify scope and becomes superfluous.  I am not a fan of `this.`

File: src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        try {
+            UserDetails userDetails = userDetailsService.getUserDetails(request.getName(), request.getCredentials());
+
+            if (userDetailsService.isValid(userDetails)) {
+                log.info("Validation of authentication response for user {} : valid", request.getName());
+                return ResponseEntity.ok(userDetails);
+            } else {
+                log.info("Validation of authentication response for user {} : invalid", request.getName());
+                return ResponseEntity.badRequest()
+                    .body(new MessageResponse(1, "Invalid data returned"));
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java]: make `1` a constant

File: src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public ResponseEntity<?> authenticateUser(AuthenticateUserRequest request) {
+
+        try {
+            UserDetails userDetails = userDetailsService.getUserDetails(request.getName(), request.getCredentials());
+
+            if (userDetailsService.isValid(userDetails)) {
+                log.info("Validation of authentication response for user {} : valid", request.getName());
+                return ResponseEntity.ok(userDetails);
+            } else {
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java]: with the `return` inside the `if` the `else` is not needed.  Drop the `else` and the `{}`

File: src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                log.info("Validation of authentication response for user {} : invalid", request.getName());
+                return ResponseEntity.badRequest()
+                    .body(new MessageResponse(1, "Invalid data returned"));
+            }
+        } catch (AuthenticationException e) {
+            MessageResponse messageResponse = new MessageResponse(1, e.getMessage());
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java]: log the request here or it'll be very difficult to find out why a given user is failing

Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/service/AuthenticateUserServiceDevKit.java]: constant for `1`

File: src/main/java/net/smartcosmos/cluster/userdetails/domain/UserDetails.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import org.hibernate.validator.constraints.NotEmpty;
+
+/**
+ * This is the response from the User Details Service that will contain the necessary
+ * information for caching purposes. While not required, if the password hash is filled
+ * this will speed up authentication considerably, since it can be queried against the
+ * native Spring Security Cache.
+ */
+@Data
+@ToString(exclude = "passwordHash")
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/domain/UserDetails.java]: exactly!

**Review by ghost — APPROVED** [review]

File: README.adoc (hunk 1 above)
Comment (andreassiegelrfid) [inline README.adoc]: Agreed. I just kept the previous message, but liked Stormpath's version more: "Invalid username or password"

File: src/main/java/net/smartcosmos/cluster/userdetails/domain/AuthenticateUserRequest.java [hunk 9]
```diff
@@ -18,7 +19,7 @@
 @NoArgsConstructor
 @Builder
 @AllArgsConstructor
-public class RestAuthenticateRequest {
+public class AuthenticateUserRequest {
```
Comment (jbauerrfid) [inline src/main/java/net/smartcosmos/cluster/userdetails/domain/AuthenticateUserRequest.java]: The DTO object should have a VERSION, like our other DTOs.

**Comment by jbauerrfid** [comment]
LGTM

**Review by jbauerrfid — APPROVED** [review]

**Pull request merged**